### ISSUE TITLE
Suggest extend over for_each for element insertion

### ISSUE THREAD

**Issue by jmg-duarte** [rust-lang/rust-clippy#8169]
### What it does

The lint should detect/convert code which uses a `.for_each` containing a single `.insert` into an `.extend` call.

### Lint Name

for_each_insert_into_extend

### Category

style

### Advantage

The code becomes much cleaner and easier to understand. 

Mentally, I read the first example as follows:
- Get all keys from `self.state_machine_info.det_states`
- For each key 
  - Insert it in `states`, *but first* `clone` the key.

The second example though:
- Extend `states`
- Using the cloned `keys` from `self.state_machine_info.det_states`

### Drawbacks

I know that not everyone will read it as I will, however the second approach feels much more in-line with "proper" Rust.

It may have performance implications, but I do not.

### Example

In my `typestate` macro I have the following code:

```rust
let mut states = HashSet::new();
self.state_machine_info.det_states.keys().for_each(|k| {
    states.insert(k.clone());
});
self.state_machine_info
    .non_det_transitions
    .keys()
    .for_each(|k| {
        states.insert(k.clone());
    });
```

Which becomes much cleaner when written like:

```rust
let mut states = HashSet::new();
states.extend(
    self.state_machine_info.det_states.keys().cloned()
);
states.extend(
    self.state_machine_info.non_det_transitions.keys().cloned()
);
```

https://github.com/rustype/typestate-rs/blob/83b0ef0a6dbd1658c283ec45be0ad611e73e8ff1/typestate-proc-macro/src/visitors/transition.rs#L99-L108

**Comment by camsteffen** [comment]
We should detect for loops to. Name idea: `manual_extend`.

**Comment by jmg-duarte** [comment]
> We should detect for loops too. Name idea: `manual_extend`.

Great suggestion! I've updated the original post!
